pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -788,26 +769,206 @@ func (r *MachineSetSyncReconciler) createOrUpdateCAPIMachineSet(ctx context.Cont
+				if capiMachineSet.Status.Conditions[j].Type == newCAPIMachineSet.Status.Conditions[i].Type {
+					hasCondition = true
+					capiMachineSet.Status.Conditions[j].Status = newCAPIMachineSet.Status.Conditions[i].Status
+					capiMachineSet.Status.Conditions[j].Reason = newCAPIMachineSet.Status.Conditions[i].Reason
+					capiMachineSet.Status.Conditions[j].Message = newCAPIMachineSet.Status.Conditions[i].Message
+
+					break
+				}
+			}
+
+			if !hasCondition {
+				capiMachineSet.Status.Conditions = append(capiMachineSet.Status.Conditions, newCAPIMachineSet.Status.Conditions[i])
+			}
+		}
+
+		// Update the v1beta2 conditions if they exist (ignoring lasttransitiontime), append new ones.
+		switch {
+		case newCAPIMachineSet.Status.V1Beta2 == nil:
+			capiMachineSet.Status.V1Beta2 = nil
+		case capiMachineSet.Status.V1Beta2 == nil:
+			capiMachineSet.Status.V1Beta2 = newCAPIMachineSet.Status.V1Beta2
+		default:
+			for i := range newCAPIMachineSet.Status.V1Beta2.Conditions {
+				hasCondition := false
+
+				for j := range capiMachineSet.Status.V1Beta2.Conditions {
+					if capiMachineSet.Status.V1Beta2.Conditions[j].Type == newCAPIMachineSet.Status.V1Beta2.Conditions[i].Type {
+						hasCondition = true
+						capiMachineSet.Status.V1Beta2.Conditions[j].Status = newCAPIMachineSet.Status.V1Beta2.Conditions[i].Status
+						capiMachineSet.Status.V1Beta2.Conditions[j].Reason = newCAPIMachineSet.Status.V1Beta2.Conditions[i].Reason
+						capiMachineSet.Status.V1Beta2.Conditions[j].Message = newCAPIMachineSet.Status.V1Beta2.Conditions[i].Message
+
+						break
+					}
+				}
+
+				if !hasCondition {
+					capiMachineSet.Status.V1Beta2.Conditions = append(capiMachineSet.Status.V1Beta2.Conditions, newCAPIMachineSet.Status.V1Beta2.Conditions[i])
+				}
+			}
+		}
+
+		if isPatchRequired, err := util.IsPatchRequired(capiMachineSet, patchBase); err != nil {
+			return fmt.Errorf("failed to check if patch is required: %w", err)
+		} else if isPatchRequired {
+			if err := r.Status().Patch(ctx, capiMachineSet, patchBase); err != nil {
+				logger.Error(err, "Failed to update CAPI machine set status")
+
+				updateErr := fmt.Errorf("failed to update CAPI machine set status: %w", err)
+
+				if condErr := r.applySynchronizedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionFalse, reasonFailedToUpdateCAPIMachineSet, updateErr.Error(), nil); condErr != nil {
+					return utilerrors.NewAggregate([]error{updateErr, condErr})
+				}
+
+				return updateErr
+			}
+
+			updated = true
+		}
+	}
+
+	if updated {
+		logger.Info("Successfully updated CAPI machine set")
+	} else {
 		logger.Info("No changes detected for CAPI machine set")
-		return nil
 	}
 
-	logger.Info("Changes detected for CAPI machine set. Updating it", "diff", fmt.Sprintf("%+v", capiMachineSetsDiff))
+	return nil
+}
+
+// createOrUpdateMAPIMachineSet creates a MAPI machine set from a CAPI one, or updates if it exists and it is out of date.
+//
+//nolint:funlen,gocognit
```

Comment:
Same as above: https://github.com/openshift/cluster-capi-operator/pull/329/files#r2330333577

Reply:
I did split it up in the most recent iterations, will push it later.